### PR TITLE
Return StatusNotFound instead of InternalServerError

### PR DESCRIPTION
**Pull request by morses-code**
Simple change to update status code returned on this specific error.

Make sure tests are still passing.

### CODE REVIEW

File: cache/elasticache_client.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+	"fmt"
+	"time"
+
+	health "github.com/ONSdigital/dp-healthcheck/healthcheck"
+	"github.com/ONSdigital/dp-sessions-api/session"
+	"github.com/go-redis/redis"
+)
+
+var (
+	HealthyMessage       = "redis is OK"
```
Comment (daiLlew) [inline cache/elasticache_client.go]: I guess this should be Elasticache not Redis.

File: cache/elasticache_client.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+// Config - config options for the redis client
+type Config struct {
+	Addr     string
+	Password string
+	Database int
+	TTL      time.Duration
+	TLS      *tls.Config
+}
+
+// New - returns new session cache instance
```
Comment (daiLlew) [inline cache/elasticache_client.go]: `// New - create a new session cache client instance`

File: cache/elasticache_client.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+type ElasticacheClient struct {
+	client RedisClienter
+	ttl    time.Duration
+}
+
+// Config - config options for the redis client
+type Config struct {
+	Addr     string
+	Password string
```
Comment (daiLlew) [inline cache/elasticache_client.go]: Being security conscientious can you add the `json:"-"` to the password field to ensure it's not accidentally logged out.

File: cache/elasticache_client.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+	sJSON, err := s.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	// Add session using ID as key
+	err = c.client.Set(s.ID, sJSON, c.ttl).Err()
+	if err != nil {
+		return fmt.Errorf("redis client.Set returned an unexpected error: %w", err)
```
Comment (daiLlew) [inline cache/elasticache_client.go]: It's not Redis anymore.

File: cache/elasticache_client.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+	}
+
+	// Add session using email as key
+	err = c.client.Set(s.Email, sJSON, c.ttl).Err()
+	if err != nil {
+		return fmt.Errorf("redis client.Set returned an unexpected error: %w", err)
```
Comment (daiLlew) [inline cache/elasticache_client.go]: ditto

File: api/handlers.go [hunk 6]
```diff
@@ -64,7 +64,7 @@ func GetByIDSessionHandlerFunc(cache Cache, getVarsFunc GetVarsFunc) http.Handle
 
 		s, err := cache.GetByID(ID)
 		if err != nil {
-			writeErrorResponse(ctx, w, err.Error(), err, getErrorStatus(err))
+			writeErrorResponse(ctx, w, err.Error(), err, http.StatusNotFound)
```
Comment (daiLlew) [inline api/handlers.go:67]: I don't think this is correct as this function can return an error for a various reasons not always "not found".

File: api/handlers.go [hunk 7]
```diff
@@ -93,7 +93,7 @@ func GetByEmailSessionHandlerFunc(cache Cache, getVarsFunc GetVarsFunc) http.Han
 
 		s, err := cache.GetByEmail(email)
 		if err != nil {
-			writeErrorResponse(ctx, w, err.Error(), err, getErrorStatus(err))
+			writeErrorResponse(ctx, w, err.Error(), err, http.StatusNotFound)
```
Comment (daiLlew) [inline api/handlers.go:96]: Same here.

File: cache/elasticache_client.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	return nil
+}
+
+// GetByID - gets a session from elasticache using its ID
+func (c *ElasticacheClient) GetByID(id string) (*session.Session, error) {
+	if id == "" {
+		return nil, ErrEmptySessionID
+	}
+
+	msg, err := c.client.Get(id).Result()
```
Comment (daiLlew) [inline cache/elasticache_client.go:97]: How does this client inform you if the requested item is not found? I think we need to specifically check for this case and return a specific error for this scenario as the caller needs to be able to say not found.

File: cache/elasticache_client.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+	err = c.Expire(s.Email, c.ttl)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// GetByEmail - gets a session from elasticache using its ID
+func (c *ElasticacheClient) GetByEmail(email string) (*session.Session, error) {
```
Comment (daiLlew) [inline cache/elasticache_client.go:125]: Same for this function.

**Review by daiLlew — CHANGES_REQUESTED** [review]

**Review by daiLlew — DISMISSED** [review]
As discussed we're going to merge this now and then follow up with some changes.

**Review by HeJones — APPROVED** [review]
The unresolved comments will be addressed in a following PR

**Pull request merged**